### ISSUE TITLE
Segmentation fault in big enough grouped convolutions

### ISSUE THREAD

**Issue by ianholing** [pytorch/pytorch#70523]
### 🐛 Describe the bug

In some environments, a big enough grouped convolution crashes with a segmentation fault. This code:

```python
import torch
from torch import nn

device = "cpu"
print("Inference device:", device)

CH = 64
x = torch.randn(1, CH, 4480, 2976)
nn.Conv2d(CH, CH, kernel_size=3, padding=1, bias=False)(x)
print ("Not grouped conv works")
nn.Conv2d(CH, CH, kernel_size=3, padding=1, groups=CH, bias=False)(x)
print ("This never shows up")
```
Executes with this output:
```
# python sigserv.py 
Inference device: cpu
Not grouped conv works
Segmentation fault (core dumped)
```

That's not true in all possible environments, I did some tests with some AWS machines with enough ram to deal with the script, the same OS image: "Deep Learning AMI (Ubuntu 18.04) Version 54.0 - ami-0954b94a3697c43dd" and the last pytorch docker (actually 1.10.0) and t2.2xlarge instances works ok but c5.2xlarge crashes.

Reproducing steps:
* Step 1
    ** Create a new instance in EC2
    ** Select this OS Image: Deep Learning AMI (Ubuntu 18.04) Version 54.0 - ami-0954b94a3697c43dd
    ** Use a c5.*xlarge as instance type
* Step 2: Inside the machine
    ** Execute: docker run -it --rm --shm-size=??gb -v $(pwd):/workspace pytorch/pytorch:latest
* Step 3:
    ** In this environment the above code will create a segmentation fault
    
Same steps changing the instance type for t2.*xlarge will work ok.

### Versions

This, in fact, is the output with the crashing machine inside the last docker image:

root@456ef378227a:/workspace# python collect_env.py
Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1060-aws-x86_64-with-debian-buster-sid
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.0
[pip3] torchelastic==0.2.0
[pip3] torchtext==0.11.0
[pip3] torchvision==0.11.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               ha36c431_9    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.3.0           h06a4308_520  
[conda] mkl-service               2.4.0            py37h7f8727e_0  
[conda] mkl_fft                   1.3.1            py37hd3c417c_0  
[conda] mkl_random                1.2.2            py37h51133e4_0  
[conda] numpy                     1.21.2           py37h20f2e39_0  
[conda] numpy-base                1.21.2           py37h79a1101_0  
[conda] pytorch                   1.10.0          py3.7_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchelastic              0.2.0                    pypi_0    pypi
[conda] torchtext                 0.11.0                     py37    pytorch
[conda] torchvision               0.11.0               py37_cu113    pytorch
root@456ef378227a:/workspace#

cc @ezyang @gchanan @zou3519 @bdhirsh @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @VitalyFedyunin

**Comment by jbschlosser** [comment]
Hey @ianholing, thanks for reporting this! Confirmed this is a problem in 1.9, 1.10, 1.10.1, and master.

Failure environment details are [here](https://gist.github.com/HamidShojanazeri/00d2b556b44ec5584306b34d6e1ed8cb) and success environment details are [here](https://gist.github.com/HamidShojanazeri/3f597776602b61e1d8263c77d2a149e9).

**Comment by ianholing** [comment]
Thanks @jbschlosser for the quick response and the workaround proposed :)

**Comment by sanchitintel** [comment]
@HamidShojanazeri & @ianholing, is it possible that T2 & C5 environment details got swapped, and that this segfault occurs on T2 AWS instances, but not on C5 (as opposed to what @ianholing reported)? Thanks!

**Comment by albanD** [comment]
> Disabling MKL-DNN results in a Killed process rather than a segfault.

This happens because you run out of RAM (most likely because our fallback algorithm is really memory hungry).

**Comment by HamidShojanazeri** [comment]
> @HamidShojanazeri & @ianholing, is it possible that T2 & C5 environment details got swapped, and that this segfault occurs on T2 AWS instances, but not on C5 (as opposed to what @ianholing reported)? Thanks!

@sanchitintel  I just double checked, it is failing on C5 and can pass on T2 instances, confirmed details [here](https://gist.github.com/HamidShojanazeri/edf0f1629adc43ec40fb1079f8e3f034).

**Comment by ianholing** [comment]
@sanchitintel maybe you are a bit confused cause if you have not enough RAM in the machine, the script will finish with a "Killed" that looks like a problem but it is a usual OOM which you could see using dmesg.

When the real bug appears, the return is a segmentation fault.

**Comment by dzarukin** [comment]
Hi @ianholing @jbschlosser!
Would it be possible to enable DNNL_VERBOSE=1 through env var, run the case and provide its output in case of segfault here, please?
That would help a lot to investigate the issue on oneDNN side. Thank you!

**Comment by jbschlosser** [comment]
> Would it be possible to enable DNNL_VERBOSE=1 through env var, run the case and provide its output in case of segfault here, please?

cc @HamidShojanazeri who has been investigating this on both instance types

**Comment by sanchitintel** [comment]
> @sanchitintel I just double checked, it is failing on C5 and can pass on T2 instances, confirmed details [here](https://gist.github.com/HamidShojanazeri/edf0f1629adc43ec40fb1079f8e3f034).

Thanks for the confirmation, @HamidShojanazeri!
I had misread the AWS instance description on the AWS website

**Comment by sanchitintel** [comment]
@ianholing, thanks for reporting this issue!

Until this issue is fixed, you can use the environment variable `DNNL_MAX_CPU_ISA=AVX2` as a workaround on systems it had been segfaulting on (i.e. you can either export the environment variable with `export DNNL_MAX_CPU_ISA=AVX2` prior to running your python script, or run the python script as `DNNL_MAX_CPU_ISA=AVX2 python script_name.py`). It'd be a bit slower, though.

**Comment by ianholing** [comment]
Great news, thank you @sanchitintel and @HamidShojanazeri, I can definitely use that workaround meanwhile.

**Comment by jbschlosser** [comment]
@sanchitintel / @dzarukin any update on this issue? has it been fixed on the oneDNN side?

**Comment by Jup11ter** [comment]
There is a similar crash `Segmentation fault (core dumped)` raised on `torch.nn.LazyConv2d` in 1.9.0 and 1.10.0.

Luckily, this bug is fixed in 1.11.0.
Since the performance of these two problems is very close, and both of them are convolution operations with `groups`, I want to confirm whether they have the same root cause.


```python

import torch
from torch import nn

device = "cpu"
print("Inference device:", device)

CH = 64
x = torch.randn(1, CH, 4480, 2976)

torch.nn.LazyConv2d(out_channels=CH, kernel_size=3, padding=1, bias=False)(x)
print ("Not grouped conv works")

torch.nn.LazyConv2d(out_channels=CH, kernel_size=3, padding=1, groups=CH, bias=False)(x)

print ("This never shows up")

```

<details>
  <summary>pytorch 1.10.0</summary>
  <pre><code>
  [pip3] numpy==1.21.2
  [pip3] torch==1.10.0
  [pip3] torch-tb-profiler==0.4.0
  [pip3] torchelastic==0.2.0
  [pip3] torchtext==0.11.0
  [pip3] torchvision==0.11.0
  [conda] blas                      1.0                         mkl  
  [conda] cudatoolkit               11.3.1               ha36c431_9    nvidia
  [conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
  [conda] mkl                       2021.3.0           h06a4308_520  
  [conda] mkl-service               2.4.0            py37h7f8727e_0  
  [conda] mkl_fft                   1.3.1            py37hd3c417c_0  
  [conda] mkl_random                1.2.2            py37h51133e4_0  
  [conda] numpy                     1.21.2           py37h20f2e39_0  
  [conda] numpy-base                1.21.2           py37h79a1101_0  
  [conda] pytorch                   1.10.0          py3.7_cuda11.3_cudnn8.2.0_0    pytorch
  [conda] pytorch-mutex             1.0                        cuda    pytorch
  [conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
  [conda] torchelastic              0.2.0                    pypi_0    pypi
  [conda] torchtext                 0.11.0                     py37    pytorch
  [conda] torchvision               0.11.0               py37_cu113    pytorch</code></pre>
</details>

<details>
  <summary>pytorch 1.11.0</summary>
  <pre><code>
  [pip3] numpy==1.21.6
  [pip3] pytorch-lightning==1.6.3
  [pip3] torch==1.11.0
  [pip3] torch-tb-profiler==0.4.0
  [pip3] torchaudio==0.11.0
  [pip3] torchmetrics==0.9.0
  [pip3] torchvision==0.12.0
  [conda] blas                      1.0                         mkl  
  [conda] cudatoolkit               11.3.1               h2bc3f7f_2  
  [conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
  [conda] mkl                       2021.4.0           h06a4308_640  
  [conda] mkl-service               2.4.0            py38h7f8727e_0  
  [conda] mkl_fft                   1.3.1            py38hd3c417c_0  
  [conda] mkl_random                1.2.2            py38h51133e4_0  
  [conda] numpy                     1.19.5                   pypi_0    pypi
  [conda] numpy-base                1.21.5           py38hf524024_1  
  [conda] pytorch                   1.11.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
  [conda] pytorch-lightning         1.6.3                    pypi_0    pypi
  [conda] pytorch-mutex             1.0                        cuda    pytorch
  [conda] torch                     1.10.0                   pypi_0    pypi
  [conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
  [conda] torchaudio                0.11.0               py38_cu113    pytorch
  [conda] torchmetrics              0.9.0                    pypi_0    pypi
  [conda] torchvision               0.12.0               py38_cu113    pytorch</code></pre>
</details>
